### PR TITLE
Fix Reflector to always close the watcher

### PR DESCRIPTION
**Pull request by karlkfi**
/bug

#### What/Why

- Ensure watcher.Stop() is called in all cases, including errors.

This is part of a multi-PR cleanup of code that uses `watch.Interface` to fix memory leaks and race conditions related to the closing of watcher stop and result channels.

Unfortunately, validating that this fix actually fixes the problem requires modifying `watch.FakeWatcher`, which is used by dozens of tests upstream. But modifying `watch.FakeWatcher` breaks the tests unless this `Reflector.ListAndWatch` bug is already fixed. So this change fixes the issue first, then I'll fix `watch.FakeWatcher` in a follow-up PR. Then I can add or modify the reflector tests to confirm that `Reflector.ListAndWatch` always closes the channels in the right order.

In a previous iteration, I had chosen to make watchHandler/handleWatch handle calling stop on the watcher, but after playing around with it a bit, I decided that doing that with defers using conditionals was confusing and error prone. So instead, I opted for ensuring that watchHandler/handleWatch never stops the watcher, instead delegating that back up to the caller, who can more easily do it with a defer without conditionals. This way, the lower level methods also do not need to call `w.Stop()` when the `stopCh` is closed.

Dependencies:
- https://github.com/kubernetes/kubernetes/pull/125299
- https://github.com/kubernetes/kubernetes/pull/125391
- https://github.com/kubernetes/kubernetes/pull/125472
- 
Related:
- https://github.com/kubernetes/kubernetes/pull/125107
- https://github.com/kubernetes/kubernetes/pull/125204
- https://github.com/kubernetes/kubernetes/pull/124801

#### Does this PR introduce a user-facing change?

```release-note
client-go: fix Reflector.ListAndWatch to always close the watcher
```

### CODE REVIEW

File: staging/src/k8s.io/client-go/tools/cache/reflector.go [hunk 1]
```diff
@@ -346,15 +346,18 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
...
 			klog.Warningf("The watchlist request ended with an error, falling back to the standard LIST/WATCH semantics because making progress is better than deadlocking, err = %v", err)
 			fallbackToList = true
 			// ensure that we won't accidentally pass some garbage down the watch.
 			w = nil
+		case w == nil:
+			// stopCh was closed
+			return nil
+		default:
+			// Ensure watcher is stopped
+			defer w.Stop()
```
Comment (wojtek-t) [inline staging/src/k8s.io/client-go/tools/cache/reflector.go:360]: This PR is mixing a bunch of refactoring with changes to the actual logic.
Can you please split those into separate PRs?

Comment (p0lyn0mial) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go:360]: Thanks, I wanted to ask for the same.

Comment (karlkfi) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go:360]: I don't think I can really fix the bug without some of the refactoring. But I can try to pull out the refactor as separate from the bug fix.

Comment (karlkfi) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go:360]: https://github.com/kubernetes/kubernetes/pull/125391

**Comment by karlkfi** [comment]
Extracted refactor to its own PR: https://github.com/kubernetes/kubernetes/pull/125391

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 1 above)
Comment (karlkfi) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go:358]: Refactoring is merged and this is rebased.

File: staging/src/k8s.io/client-go/tools/cache/reflector.go [hunk 2]
```diff
@@ -346,15 +346,16 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
...
 			klog.Warningf("The watchlist request ended with an error, falling back to the standard LIST/WATCH semantics because making progress is better than deadlocking, err = %v", err)
 			fallbackToList = true
-			// ensure that we won't accidentally pass some garbage down the watch.
-			w = nil
+		case w == nil:
+			// stopCh was closed, so no error or watcher was returned
+			return nil
+		default:
+			// Ensure the watcher is always stopped on exit, including error or stopCh closure.
+			defer w.Stop()
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/client-go/tools/cache/reflector.go:358]: won't `w.Stop()` already be stopped in https://github.com/kubernetes/kubernetes/blob/c7c422da8bb340d4fe4cc36a48a6b4751269fe91/staging/src/k8s.io/client-go/tools/cache/reflector.go#L468 ?

Comment (karlkfi) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go:358]: Only if it makes it that far. 
Theres 2 or 3 shopCh checks before then that return early, and the code is complex enough it's better to be defensive. And in case of panic the defer still runs, though that's a lesser concern since the system's probably already hosed on panic.

File: staging/src/k8s.io/client-go/tools/cache/reflector.go [hunk 3]
```diff
@@ -687,8 +706,14 @@ func (r *Reflector) watchList(stopCh <-chan struct{}) (watch.Interface, error) {
...
 		}
 		if watchListBookmarkReceived {
+			// If we go the watchlist bookmark, stop consuming events,
+			// but do not stop the watcher.
 			break
 		}
+		// It's possible that the producer closed the result channel before the
+		// watchlist bookmark was received, without first sending an error
+		// event. If so, stop the watcher.
+		w.Stop()
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/client-go/tools/cache/reflector.go:716]: okay, i think this is the following case https://github.com/kubernetes/kubernetes/blob/c7c422da8bb340d4fe4cc36a48a6b4751269fe91/staging/src/k8s.io/client-go/tools/cache/reflector.go#L829

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 2 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go:358]: can we ensure that calling `w.Stop` more than once won't panic ?

Comment (wojtek-t) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go:358]: I'm not a fan of the pattern of "it's so complex that worst case close it twice". We don't necessary control all implementations and closing it more than once can lead to problems.

We should rather introduce a contract that is somewhat maintainable and meet the contract.

It seems that the following seems to be relatively simple to achieve:
`If w is non-nil, then w is not stopped`

With the above contract, we can switch here to the following:
```
defer func() { if w != nil {  w.Stop() } }
```

and it should now work fine [we should also update line 699 and set `w=nil` there probably]

Comment (karlkfi) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go:358]: All of the watch.Interface implementations I've seen either use a mutex and a stopped bool OR a sync.Once to prevent closing their internal stop channel more than once when Stop is called. This is also how the CancelFunc works from context.WithCancel, http response body, test cleanup, etc. It's a pretty common pattern, because it allows starting something and immediately creating a defer to stop it, even if the context or watcher gets passed through multiple subsequent functions which may also call stop.

I think it would be fine to add that expectation as a comment on the Stop method of watch.Interface. 

The reason I opted for calling stop multiple times here is simply to avoid needing another refactor.

I tried using `defer func() { if w != nil {  w.Stop() } }` initially, without refactoring out inner functions, but it's problematic in some cases. For example, in `watchList`, we only want to stop the watcher if there was an error or the result channel was closed before the bookmark was received. If the bookmark was received, we want to leave it open for the subsequent `watch` to consume.

And in `watch`, the `w` argument is being modified to point to a new watch impl on every loop iteration. So there's a subtle difference between `defer w.Stop()` and `defer func() { if w != nil {  w.Stop() } }`, in that the former always only stops the initial watcher (which may already have been stopped or may be nil) and the later only stops the last watcher.

So we could use `defer func() { if w != nil {  w.Stop() } }` in `watch` but not in `watchList`. A `watchList` defer has to keep track of whether the watcher needs to stay open or not, which requires checking a few variables. Doing that seemed inconsistent and verbose, so I opted add the defer at the top level and make as few changes as possible to `watch` and `watchlist`.

If you want though, I can take another stab at extracting out the iteration code as functions in one PR and then make yet another PR to add the deferred stops.

Comment (karlkfi) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go:358]: As I mentioned elsewhere, the watch.Interface is not ideal. The methods don't make it obvious how you're supposed to use or implement it.

If we could change that, I would just make a single ResultChan method that takes a stop channel or Context as an argument, with no Stop method. That way it would be obvious that it's the consumer's responsibility to close the stop channel or cancel/timeout the context, and they can do it however they please, without needing to known how the producer implemented the Stop method.

But I figured changing that was too big of a change, since it would break just about every controller, tool, and K8s integration.

File: staging/src/k8s.io/client-go/tools/cache/reflector.go (hunk 3 above)
Comment (karlkfi) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go:716]: Yes. We could make that throw an error, but that seemed like a behavioral change I didn't want to introduce without knowing the consequences to callers.

Comment (karlkfi) [reply staging/src/k8s.io/client-go/tools/cache/reflector.go:716]: FWIW, most implementations only close the result channel after an error event or after the stop channel is closed, but there's not really any way to enforce that. Doing so in a ResultChan comment wont fix any impls that already have this bug. So I figured we need to handle it defensively here.

**Comment by karlkfi** [comment]
Replaced by https://github.com/kubernetes/kubernetes/pull/131266
